daft/sql/sql_scan.py

```diff
@@ -33,14 +36,27 @@ class SQLScanOperator(ScanOperator):
     def __init__(
         self,
         sql: str,
-        url: str,
+        conn: str | Callable[[], Connection],
         storage_config: StorageConfig,
         partition_col: str | None = None,
         num_partitions: int | None = None,
     ) -> None:
         super().__init__()
         self.sql = sql
-        self.url = url
+        self.conn = conn
```

Comment:
WDYT if instead of storing `self.conn: Callable | str`, we store a `self.conn: ConnectionFactory` and we can create a wrapper class for wrapping the incoming user-provided factory method, or user-provided string?

```
self.conn = if isinstance(conn, str) ConnectionFactory.from_url(conn) else ConnectionFactory.from_user_func(conn)
```

This will help us remove lots of `ifinstance(self.conn, str)` in the rest of the code in this PR, and we can more closely control serialization of this factory.

Reply:
Yup totally agree, and also it will make things a lot easier if we introduce ADBC connections